### PR TITLE
Refactor Canvas Post-Processing logic in CanvasBase

### PR DESCRIPTION
**Pull request by sysrqb**
#### 44522d1c3a43d38639df8760d6be6ad964470f46
<pre>
Refactor Canvas Post-Processing logic in CanvasBase
<a href="https://bugs.webkit.org/show_bug.cgi?id=255756">https://bugs.webkit.org/show_bug.cgi?id=255756</a>
rdar://106872464

Reviewed by Kimmo Kinnunen.

When the noise injection policy is active, we are currently post-processing
Canvas2D pixel buffers, but that processing was not applied to WebGL image data
when alpha was premultiplied. This change fixes that bug by refactoring the
logic and moving it into CanvasBase. For Canvas2D, this patch tracks the dirty
rect from each drawing operation and only that area is post-processed before
Canvas image data is implicitly extracted (e.g., toDataURL). For WebGL, the
entire Canvas&apos; image data is post-processed.

This patch leaves tracking CanvasRenderingContext2DBase::putImageData rects as
future work because those rects do not need post-processing.

* Source/WebCore/html/CanvasBase.cpp:
(WebCore::CanvasBase::makeRenderingResultsAvailable):
(WebCore::CanvasBase::didDraw): Here we now track the dirty area if we should
post-process the data in the future.

(WebCore::CanvasBase::shouldInjectNoiseBeforeReadback const):

(WebCore::CanvasBase::postProcessDirtyCanvasBuffer const):
(WebCore::CanvasBase::postProcessPixelBufferResults const):
(WebCore::CanvasBase::postProcessPixelBuffer const): Deleted.
These functions are moved and renamed from CanvasRenderingContext2DBase.

* Source/WebCore/html/CanvasBase.h:
(WebCore::CanvasBase::didDraw):
* Source/WebCore/html/CustomPaintCanvas.h:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::didDraw):
(WebCore::HTMLCanvasElement::getImageData):
(WebCore::HTMLCanvasElement::setImageBufferAndMarkDirty):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::didDraw):
(WebCore::OffscreenCanvas::setImageBufferAndMarkDirty):
* Source/WebCore/html/OffscreenCanvas.h:
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::postProcessPixelBuffer const): Deleted.
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::computeImageDataRect):
(WebCore::CanvasRenderingContext2DBase::didDrawEntireCanvas):
(WebCore::CanvasRenderingContext2DBase::didDraw):
(WebCore::CanvasRenderingContext2DBase::getImageData const): Now this
function behaves in the same way as the other APIs that extract image data.

(WebCore::CanvasRenderingContext2DBase::putImageData):

(WebCore::CanvasRenderingContext2DBase::postProcessPixelBuffer const): Deleted.
This function is moved into CanvasBase.

* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
(WebCore::CanvasRenderingContext2DBase::didDraw):

Canonical link: <a href="https://commits.webkit.org/263694@main">https://commits.webkit.org/263694@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/f95ae973aea13210167ad714377febc92139697d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5442 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5583 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5776 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6985 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5469 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5442 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5811 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5560 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/6417 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5544 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5605 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4874 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7014 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3083 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4881 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/12029 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4949 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4961 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6675 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5396 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/4429 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4856 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1304 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8958 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5219 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp [hunk 1]
```diff
@@ -3560,6 +3560,19 @@ IntRect WebGLRenderingContextBase::getImageDataSize(ImageData* pixels)
...
+void WebGLRenderingContextBase::postProcessPixelBuffer() const
+{
+    if (!canvasBase().shouldInjectNoiseBeforeReadback())
+        return;
+
+    ImageBuffer* buffer = canvasBase().buffer();
+    if (!buffer)
+        return;
+
+    auto canvasRect = texImageSourceSize(canvasBase());
```
Comment (kkinnunen-apple) [inline Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp]: The function 'tezImageSoruceSize' is incorrectly used. E.g. if you have a canvasBase, you know how to get its size without any additional function.
texImageSourceSize function is intended to be used for TexImage sources  for arbitrary source T. The canvas here isn't used as texImageSource, it's just this canvas.

As you can observe, there's nothing webgl about this function. The logic of this function is:
   Post process each pixel that changed after previous post process.

This applies to any canvas.

So one option would be that you would make following changes:
- Move CanvasRenderingContext::postProcessPixelBuffer() to CanvasBase::postProcessPixelBuffer()
- Rename  CanvasBase::postProcessPixelBuffer() to  CanvasBase::postProcessRenderingResults()
- Move 
 CanvasRenderingContext::postProcessPixelBuffer(ImageBuffer&, IntRect, const HashSet<uint32_t>&) to CanvasBase::postProcessPixelBuffer(ImageBuffer&, IntRect, const HashSet<uint32_t>&)
- Introduce CanvasBase::didDraw(const std::optional<FloatRect>& rect, SomeOptions) where you record m_postProcessDirtyRect
- Remove m_wasLastDrawPutImageData, this is superseeded by m_postProcessDirtyRect. Current m_wasLastDrawPutImageData doesn't seem to handle putImageData correctly, it will allow attacker to circumvent the post processing by calling with for example 1,1 region as per my reading of my checkout, I may be wrong though
- Make SomeOption to be able to get "SomeOption::TheDrawWasActuallyPutImageDataSoDoNotAccumulateThisToPostProcessingDirtyRect"



So in other words:
- canvas drawing contexts signal CanvasBase that "something changed, maintain the post processing state"
- CanvasBase already has the image buffer for the display buffer  of the context. CanvasBase is already the sole entity giving the pixels out, with the exception of WebGL non-premul alpha PixelBuffer case. Make CanvasBase apply the post processing.

Comment (sysrqb) [reply Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp]: Thanks for the review! I agree that your suggestion is better.

File: Source/WebCore/html/CanvasBase.h [hunk 2]
```diff
@@ -127,7 +130,10 @@ class CanvasBase {
     virtual void dispatchEvent(Event&) = 0;
 
     bool shouldInjectNoiseBeforeReadback() const;
-    bool postProcessPixelBuffer(Ref<PixelBuffer>&&, bool, const HashSet<uint32_t>&) const;
+    void postProcessDirtyCanvasBuffer() const;
```
Comment (kkinnunen-apple) [inline Source/WebCore/html/CanvasBase.h]: You could check which of these functions can be moved as private now.

File: Source/WebCore/html/CanvasBase.h [hunk 3]
```diff
@@ -109,7 +112,7 @@ class CanvasBase {
 
     GraphicsClient* graphicsClient() const;
 
-    virtual void didDraw(const std::optional<FloatRect>&) = 0;
+    virtual void didDraw(const std::optional<FloatRect>&, const ShouldApplyPostProcessingToDirtyRect = ShouldApplyPostProcessingToDirtyRect::Yes);
```
Comment (kkinnunen-apple) [inline Source/WebCore/html/CanvasBase.h]: In normal C++, the default arguments for virtual functions are frowned upon, as they're hard to keep in sync.
typical suggestion is to add inline overload for the default call, in your case
```
void didDraw(const std::optional<FloatRect> rect)  ( didDraw(rect, ...DirtyRect::Yes };
virtual void didDraw(const std::optional<FloatRect>&, ShouldApplyPostProcessingToDirtyRect);
```

This way you don't need the default argument to match in all the overrides.

(Note, WebKit C++ does have default parameters in the virtual functions in many places. Still, I'd argue adding more of these wouldn't be such a good idea)

File: Source/WebCore/html/HTMLCanvasElement.cpp [hunk 4]
```diff
@@ -558,6 +558,8 @@ void HTMLCanvasElement::didDraw(const std::optional<FloatRect>& rect)
         return;
     }
 
+    CanvasBase::didDraw(rect, shouldApplyPostProcessingToDirtyRect);
```
Comment (kkinnunen-apple) [inline Source/WebCore/html/HTMLCanvasElement.cpp]: I think it would be consistent to forward the base class call first, and then do whatever logic the subclass needs.

File: Source/WebCore/html/OffscreenCanvas.cpp [hunk 5]
```diff
@@ -420,7 +420,7 @@ void OffscreenCanvas::convertToBlob(ImageEncodeOptions&& options, Ref<DeferredPr
     promise->resolveWithNewlyCreated<IDLInterface<Blob>>(WTFMove(blob));
 }
 
-void OffscreenCanvas::didDraw(const std::optional<FloatRect>& rect)
+void OffscreenCanvas::didDraw(const std::optional<FloatRect>& rect, const ShouldApplyPostProcessingToDirtyRect)
 {
     clearCopiedImage();
```
Comment (kkinnunen-apple) [inline Source/WebCore/html/OffscreenCanvas.cpp:426]: I think here you would still forward the call to the base class.

File: Source/WebCore/html/CanvasBase.cpp [hunk 6]
```diff
@@ -199,6 +200,12 @@ void CanvasBase::notifyObserversCanvasChanged(const std::optional<FloatRect>& re
         observer.canvasChanged(*this, rect);
 }
 
+void CanvasBase::didDraw(const std::optional<FloatRect>& rect, const ShouldApplyPostProcessingToDirtyRect shouldApplyPostProcessingToDirtyRect)
+{
+    if (shouldInjectNoiseBeforeReadback() && shouldApplyPostProcessingToDirtyRect == ShouldApplyPostProcessingToDirtyRect::Yes && rect && !m_postProcessDirtyRect.contains(*rect))
```
Comment (kkinnunen-apple) [inline Source/WebCore/html/CanvasBase.cpp]: `if (!a.contains(b)) a.unite(b);` == `a.unite(b);`
I.e. no need to check `contains`, at that is equal to just doing the `unite` in terms of complexity

I think it would be better to categorize the data in rect as "something needing postprocess" and "something does not need post process". 

So we have a goal to give out prestine putImageData back in some cases. 

With your patch, consider:
 empty canvas + full image putImageData -> you can get prestine putImageData back
 empty canvas + draw line + full image putimageData -> you cannot get prestine putImageData back

It is not logical. If putImageData is "not need postprocess", you should always get prestine putImageData back in relevant, same cases.

File: Source/WebCore/html/CanvasBase.cpp [hunk 7]
```diff
@@ -361,12 +368,50 @@ RefPtr<ImageBuffer> CanvasBase::allocateImageBuffer(bool usesDisplayListDrawing,
 
 bool CanvasBase::shouldInjectNoiseBeforeReadback() const
 {
+    // Note, every early-return resulting from this check potentially leaks this state. This is a risk that we're accepting right now.
     return scriptExecutionContext() && scriptExecutionContext()->noiseInjectionHashSalt();
 }
 
-bool CanvasBase::postProcessPixelBuffer(Ref<PixelBuffer>&& pixelBuffer, bool wasLastDrawByBitMap, const HashSet<uint32_t>& suppliedColors) const
+IntRect CanvasBase::computeImageDataRect(const ImageBuffer& buffer, int width, int height, IntRect destRect, const IntSize& destOffset)
```
Comment (kkinnunen-apple) [inline Source/WebCore/html/CanvasBase.cpp]: This function is relevant only when computing the source rect used in getImageData() Context2D call. It's an odd function used for odd reason (and implemented in a bit odd way), so it should stay at its odd place, at Context2D.



Here in CanvasBase we are concerned about the imagebuffer, and there is nothing mysterious about its sizes.

File: Source/WebCore/html/CanvasBase.cpp [hunk 8]
```diff
@@ -361,12 +368,50 @@ RefPtr<ImageBuffer> CanvasBase::allocateImageBuffer(bool usesDisplayListDrawing,
...
+    auto imageRect = !m_postProcessDirtyRect.isEmpty() ? enclosingIntRect(m_postProcessDirtyRect) : IntRect { { }, size() };
+    auto colorSpace = renderingContext() ? renderingContext()->colorSpace() : DestinationColorSpace::SRGB();
+    PixelBufferFormat format { AlphaPremultiplication::Unpremultiplied, PixelFormat::RGBA8, colorSpace };
+    auto pixelBuffer = imageBuffer->getPixelBuffer(format, imageRect);
+    if (!is<ByteArrayPixelBuffer>(pixelBuffer))
+        return;
+
+    if (postProcessPixelBufferResults(*pixelBuffer, { }, imageRect)) {
+        IntSize destOffset { 0, 0 };
+        IntRect sourceRect = computeImageDataRect(*imageBuffer, imageRect.width(), imageRect.height(), imageRect, destOffset);
```
Comment (kkinnunen-apple) [inline Source/WebCore/html/CanvasBase.cpp]: So here it's pointless to use an helper function to compute the source rect.
Source rect is the dirty postprocess rect.
Destination rect is the dirty postprocess rect.

File: Source/WebCore/html/CanvasBase.cpp [hunk 9]
```diff
@@ -361,12 +368,50 @@ RefPtr<ImageBuffer> CanvasBase::allocateImageBuffer(bool usesDisplayListDrawing,
...
+void CanvasBase::postProcessDirtyCanvasBuffer() const
+{
+    if (!shouldInjectNoiseBeforeReadback())
+        return;
+
+    ImageBuffer* imageBuffer = buffer();
+    if (!imageBuffer)
+        return;
+
+    auto imageRect = !m_postProcessDirtyRect.isEmpty() ? enclosingIntRect(m_postProcessDirtyRect) : IntRect { { }, size() };
```
Comment (kkinnunen-apple) [inline Source/WebCore/html/CanvasBase.cpp]: if the post process dirty rect is empty, it should be logical that there's nothing to postprocess.

File: Source/WebCore/html/CanvasBase.cpp [hunk 10]
```diff
@@ -361,12 +368,50 @@ RefPtr<ImageBuffer> CanvasBase::allocateImageBuffer(bool usesDisplayListDrawing,
...
+    auto colorSpace = renderingContext() ? renderingContext()->colorSpace() : DestinationColorSpace::SRGB();
```
Comment (kkinnunen-apple) [inline Source/WebCore/html/CanvasBase.cpp]: Here we are postprocessing the image buffer attached to the canvas element.
as such, the image buffer itself has a color space, and we shouldn't need to consult the possible context2d for the colorspace.

File: Source/WebCore/html/CanvasBase.cpp [hunk 11]
```diff
@@ -361,12 +368,50 @@ RefPtr<ImageBuffer> CanvasBase::allocateImageBuffer(bool usesDisplayListDrawing,
...
+        IntRect sourceRect = computeImageDataRect(*imageBuffer, imageRect.width(), imageRect.height(), imageRect, destOffset);
+        imageBuffer->putPixelBuffer(*pixelBuffer, sourceRect, IntPoint { destOffset });
+        m_postProcessDirtyRect = { };
+    }
+}
+
+bool CanvasBase::postProcessPixelBufferResults(Ref<PixelBuffer>&& pixelBuffer, const HashSet<uint32_t>& suppliedColors, const IntRect& imageRect) const
 {
-    if (!shouldInjectNoiseBeforeReadback() || wasLastDrawByBitMap)
+    if (!shouldInjectNoiseBeforeReadback() || (!m_postProcessDirtyRect.isEmpty() && !m_postProcessDirtyRect.intersects(imageRect)))
```
Comment (kkinnunen-apple) [inline Source/WebCore/html/CanvasBase.cpp]: the caller resets the postprocess dirty rect. This means it should be logical that the caller also consults the post process dirty rect.

Suppose caller calls toData(), inducing a post process for the image buffer.
Now you have a semi-bug where the postProcessPixelBufferResults is called by the context2d in getImageData case, and it subtly happens to not do anything, as expected, because the toData did reset the dirty rect. However, the same thing doesn't work for webgl non-alpha case, where the toData would reset the dirty rect, and then the postProcessPixelBufferResults would early return incorrectly and unexpectedly.

instead, postProcessPixelBufferResults should mostly be internal business of the CanvasBase, to the degree it is able to be.

File: Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp [hunk 12]
```diff
@@ -2368,11 +2332,7 @@ ExceptionOr<Ref<ImageData>> CanvasRenderingContext2DBase::getImageData(int sx, i
 
     ASSERT(pixelBuffer->format().colorSpace == toDestinationColorSpace(computedColorSpace));
 
-    if (canvasBase().postProcessPixelBuffer(*pixelBuffer, m_wasLastDrawPutImageData, suppliedColors())) {
-        IntSize destOffset { 0, 0 };
-        IntRect sourceRect = computeImageDataRect(*buffer, imageDataRect.width(), imageDataRect.height(), imageDataRect, destOffset);
-        buffer->putPixelBuffer(*pixelBuffer, sourceRect, IntPoint { destOffset });
-    }
+    canvasBase().postProcessPixelBufferResults(*pixelBuffer, suppliedColors(), imageDataRect);
```
Comment (kkinnunen-apple) [inline Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp]: This call seems like a layering violation, still.
I'd consider it a better approach if above, you had:
```
canvasBase().makeRenderingResultsAvailable();
ImageBuffer* buffer = canvasBase().buffer();
````

In other words: the post processing would be always maintained before an escaping read.

**Review by kkinnunen-apple — COMMENTED** [review]
Maybe at least the smaller concerns could be fixed before landing

Also, I think this is testable with layout tests, so it'd be great to see some layout tests in some time in the future.

File: Source/WebCore/html/CanvasBase.cpp (hunk 6 above)
Comment (sysrqb) [reply Source/WebCore/html/CanvasBase.cpp]: > if (!a.contains(b)) a.unite(b); == a.unite(b);

Good point.

> So we have a goal to give out prestine putImageData back in some cases.

I was approaching this as best-effort because I made a wrong assumption that `alpha != 1` would cause compositing. That is obviously wrong in retrospect.

> If putImageData is "not need postprocess", you should always get prestine putImageData back in relevant, same cases.

I agree this is the ideal and expected behavior, but I'll add this in a follow-up patch. Supporting this makes the implementation more complicated.

File: Source/WebCore/html/CanvasBase.cpp (hunk 7 above)
Comment (sysrqb) [reply Source/WebCore/html/CanvasBase.cpp]: I factored out that code in 1a169d095a6691b2461810e3b639fffed1cf87f6, so we can restore the previous code. I'll do that in a follow-up patch.

File: Source/WebCore/html/CanvasBase.cpp (hunk 8 above)
Comment (sysrqb) [reply Source/WebCore/html/CanvasBase.cpp]: Good point.

File: Source/WebCore/html/CanvasBase.cpp (hunk 9 above)
Comment (sysrqb) [reply Source/WebCore/html/CanvasBase.cpp]: This was another wrong assumption that WebGL didn't emit didDraw - but it does, so indeed this was wrong. This is now changed.

File: Source/WebCore/html/CanvasBase.cpp (hunk 10 above)
Comment (sysrqb) [reply Source/WebCore/html/CanvasBase.cpp]: Corrected, thanks for noticing that.

File: Source/WebCore/html/CanvasBase.cpp (hunk 11 above)
Comment (sysrqb) [reply Source/WebCore/html/CanvasBase.cpp]: That's a good suggestion.

File: Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp (hunk 12 above)
Comment (sysrqb) [reply Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp]: That's a good suggestion. The one reason I am using `postProcessPixelBufferResults` here is because this is a case where the "post processing" is not written back to the ImageBuffer and it does not result in repainting the canvas with the changes. Ideally, post-processing should only affect the script-readable extracted data. Currently, this isn't true because we rely on `makeRenderingResultsAvailable` for handling the other extraction methods. I see your point that this is a layering violation, and maybe this isn't sufficiently valuable.

File: Source/WebCore/html/CanvasBase.h [hunk 13]
```diff
@@ -109,7 +112,8 @@ class CanvasBase {
 
     GraphicsClient* graphicsClient() const;
 
-    virtual void didDraw(const std::optional<FloatRect>&) = 0;
+    void didDraw(const std::optional<FloatRect>& rect) { return didDraw(rect, ShouldApplyPostProcessingToDirtyRect::Yes); }
+    virtual void didDraw(const std::optional<FloatRect>&, const ShouldApplyPostProcessingToDirtyRect);
```
Comment (kkinnunen-apple) [inline Source/WebCore/html/CanvasBase.h]: Forgot to mention this before: const of by-value function parameters are not useful, so you could remove the const from ShouldApplyPostProcessingToDirtyRect.

**Review by kkinnunen-apple — APPROVED** [review]

File: Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp (hunk 12 above)
Comment (kkinnunen-apple) [reply Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp]: > Ideally, post-processing should only affect the script-readable extracted data.

I think this is a really good idea, but the code has to realize this, then, ideally for all cases. I think mix-and-match sometimes mutating sometimes not is a bad combination.
